qa/common/infer_util.py

```diff
@@ -400,3 +437,445 @@ def infer_exact(tester, pf, tensor_shape, batch_size,
+                          priority=priority, timeout_us=timeout_us)
+
+        tester.assertEqual(ctx.get_last_request_model_name(), model_name)
+        if model_version is not None:
+            tester.assertEqual(ctx.get_last_request_model_version(), model_version)
+
+        tester.assertEqual(len(results), io_cnt)
+        for (result_name, result_val) in iteritems(results):
+            tester.assertTrue(result_name in output_dict)
+            tester.assertTrue(result_name in expected_dict)
+            for b in range(batch_size):
+                expected = expected_dict[result_name][b]
+                tester.assertEqual(result_val[b].shape, expected.shape)
+                tester.assertTrue(np.array_equal(result_val[b], expected),
+                                  "{}, {}, slot {}, expected: {}, got {}".format(
+                                      model_name, result_name, b, expected, result_val[b]))
+
+    if len(shm_ip_handles) != 0:
+        for io_num in range(io_cnt):
+            shared_memory_ctx.unregister(shm_ip_handles[io_num])
+            shared_memory_ctx.unregister(shm_op_handles[io_num])
+            su.destroy_either_shm_region(shm_ip_handles[io_num], use_system_shared_memory, use_cuda_shared_memory)
+            su.destroy_either_shm_region(shm_op_handles[io_num], use_system_shared_memory, use_cuda_shared_memory)
+
+    return results
+
+
+# Perform inference on a model that takes a shape and a dummy tensors as inputs,
+# resize the dummy tensor with the provided values in the shape tensor and finally
+# return the shape of the resized tensor.
+def infer_shape_tensor(tester, pf, batch_size, tensor_dtype, input_shape_values, dummy_input_shapes,
+                       model_version=None, use_http=True, use_grpc=True, use_http_json_tensors=True,
+                       use_streaming=True, shm_suffix="", use_system_shared_memory=False,
+                       use_cuda_shared_memory=False, priority=0, timeout_us=0):
+    tester.assertTrue(
+        use_http or use_http_json_tensors or use_grpc or use_streaming)
+    configs = []
+    if use_http:
+        configs.append(("localhost:8000", "http", False, True))
+    if use_http_json_tensors and (tensor_dtype != np.float16):
+        configs.append(("localhost:8000", "http", False, False))
+    if use_grpc:
+        configs.append(("localhost:8001", "grpc", False, False))
+    if use_streaming:
+        configs.append(("localhost:8001", "grpc", True, False))
+    tester.assertEqual(len(input_shape_values), len(dummy_input_shapes))
+    io_cnt = len(input_shape_values)
+
+    if use_system_shared_memory and use_cuda_shared_memory:
+        raise ValueError(
+            "Cannot set both System and CUDA shared memory flags to 1")
+
+    input_dict = {}
+    # output_dict = {}
+    expected_dict = {}
+    shm_ip_handles = list()
+    shm_op_handles = list()
+
+    shm_client = httpclient.InferenceServerClient("localhost:8000")
```

Comment:
You are using the http endpoint only for registering the shared memory. Can you use something like:

>  if config[1] == "http":
>             triton_client = httpclient.InferenceServerClient(
>                 config[0], verbose=True)
>         else:
>             triton_client = grpcclient.InferenceServerClient(
>                 config[0], verbose=True)

and then use triton_client to set up the shared memory and send the request? No need of having separate triton_client and shm_client. This way we can improve the test coverage.

Reply:
This is created once for all the configs

Reply:
That will limit the test environment to always have HTTP client installed, while the declaration of the test util suggests that it can be used to test HTTP / GRPC independently. i.e. we can't just install GRPC client and use the test for GRPC only.

Reply:
Can you can create the shared memory region once, but inside the loop, register and unregister these precreated shared memory regions using appropriate client? Otherwise we are missing out on gRPC testing.

Reply:
Fixed